generalised fee methods

ExtremelySunnyYK opened the pull request.

jmank88 commented on common/fee/models.go:
```diff
@@ -0,0 +1,183 @@
+	return bumpedFeePrice, nil
+}
+
+func getMaxFeePrice(userSpecifiedMax, maxFeePrice *big.Int) *big.Int {
+	return FeePriceLimit(userSpecifiedMax, maxFeePrice)
```
Could the callers just call `FeePriceLimit` directly instead?

ExtremelySunnyYK replied:
This would reduce the clutter from unnecessary abstraction, thanks for the catch!

ExtremelySunnyYK replied:
was initally thinking of having it there because of the naming, but `getMaxFeePrice` does not represent the actual logic of the function which is finding the min max

jmank88 commented on common/fee/models.go:
```diff
@@ -0,0 +1,183 @@
+	return FeePriceLimit(userSpecifiedMax, maxFeePrice)
+}
+
+func bumpFeePriceByPercentage(originalFeePrice *big.Int, feeBumpPercent uint16, feeBumpUnits *big.Int) *big.Int {
+	return max(new(big.Int).Add(originalFeePrice, feeBumpUnits), addPercentage(originalFeePrice, feeBumpPercent))
```
Comment? Or maybe just reformat so it reads more plainly?
```suggestion
func bumpFeePriceByPercentage(originalFeePrice *big.Int, feeBumpPercent uint16, feeBumpUnits *big.Int) *big.Int {
	return max(
		new(big.Int).Add(originalFeePrice, feeBumpUnits), 
		addPercentage(originalFeePrice, feeBumpPercent),
	)
```

ExtremelySunnyYK replied:
thanks for the reformat. It looks much cleaner

jmank88 commented on common/fee/models.go:
```diff
@@ -0,0 +1,183 @@
+}
+
+func maxBumpedFee(lggr logger.SugaredLogger, currentFeePrice, bumpedFeePrice, maxFeePrice *big.Int, feeType string) *big.Int {
+	if currentFeePrice != nil {
+		if currentFeePrice.Cmp(maxFeePrice) > 0 {
+			// Shouldn't happen because the estimator should not be allowed to
+			// estimate a higher fee than the maximum allowed
+			lggr.AssumptionViolationf("Ignoring current %s of %s that would exceed max %s of %s", feeType, currentFeePrice.String(), feeType, maxFeePrice.String())
+		} else if bumpedFeePrice.Cmp(currentFeePrice) < 0 {
+			// If the current fee price is higher than the old price bumped, use that instead
+			bumpedFeePrice = currentFeePrice
+		}
+	}
+	return bumpedFeePrice
```
Could return early to reduce indentation. Usually easier to reason about in this format too :shrug: 
Also, would it be simpler to just return the `currentFeePrice` directly?
```suggestion
	if currentFeePrice == nil {
		return bumpedFeePrice
	}
	if currentFeePrice.Cmp(maxFeePrice) > 0 {
		// Shouldn't happen because the estimator should not be allowed to
		// estimate a higher fee than the maximum allowed
		lggr.AssumptionViolationf("Ignoring current %s of %s that would exceed max %s of %s", feeType, currentFeePrice.String(), feeType, maxFeePrice.String())
	} else if bumpedFeePrice.Cmp(currentFeePrice) < 0 {
		// If the current fee price is higher than the old price bumped, use that instead
		return currentFeePrice
	}
	return bumpedFeePrice
```

ExtremelySunnyYK replied:
that's much more readable. Thanks for the suggestion

jmank88 commented on common/fee/models.go:
```diff
@@ -0,0 +1,183 @@
+	return feeCap
+}
+
+// BumpDynamicFeeOnly bumps the tip cap and max fee price if necessary
+func CalculateBumpDynamicFee(cfg feetypes.BumpConfig, feeCapBufferBlocks uint16, lggr logger.SugaredLogger, currentTipCap, currentBaseFee *big.Int, originalFeeCap, originalTipCap *big.Int, originalFeeLimit uint32, maxFeePrice *big.Int) (bumpedFeeCap, bumpedTipCap *big.Int, chainSpecificFeeLimit uint32, err error) {
+	bumpedFeeCap, bumpedTipCap, err = bumpDynamicFee(cfg, feeCapBufferBlocks, lggr, currentTipCap, currentBaseFee, originalFeeCap, originalTipCap, maxFeePrice)
+	if err != nil {
+		return bumpedFeeCap, bumpedTipCap, 0, err
+	}
+	chainSpecificFeeLimit = ApplyMultiplier(originalFeeLimit, cfg.LimitMultiplier())
+	return
```
This func looks like it is doing two separate things - can we separate these? Or maybe just inline them at that point? The callers already have to handle the error anyways, so I think it will be simpler to read.
Old:
```go
bumpedFeeCap, bumpedTipCap, chainSpecificGasLimit, err :=
		commonfee.CalculateBumpDynamicFee(f.config, f.bhConfig.EIP1559FeeCapBufferBlocks(),
			f.lggr, f.config.TipCapDefault(),
			nil, originalFee.FeeCap.ToInt(), originalFee.TipCap.ToInt(),
			originalGasLimit, maxGasPriceWei.ToInt())
	if err != nil {
		return DynamicFee{}, 0, err
	}
```
New:
```go
bumpedFeeCap, bumpedTipCap, err :=
		commonfee.BumpDynamicFee(f.config, f.bhConfig.EIP1559FeeCapBufferBlocks(),
			f.lggr, f.config.TipCapDefault(), nil, originalFee.FeeCap.ToInt(), 
			originalFee.TipCap.ToInt(), maxGasPriceWei.ToInt())
	if err != nil {
		return DynamicFee{}, 0, err
	}
	chainSpecificFeeLimit = commonfee.ApplyMultiplier(originalGasLimit, f.config.LimitMultiplier())
```

ExtremelySunnyYK replied:
ahh got it. `chainSpecificFeeLimit = ApplyMultiplier(originalFeeLimit, cfg.LimitMultiplier())` Yeah this part seems to be pretty confusing. i think that would be a good idea to seperate them as well.

ExtremelySunnyYK replied:
when i was reading the code, the part seemed to confuse me a little. Will extract out chain specific gas limit. Thanks for the note Jordan!

jmank88 commented on common/fee/models.go:
```diff
@@ -0,0 +1,183 @@
+
+	// Always bump the FeeCap by at least the bump percentage
+	// For geth the configured bump is 10%
+	bumpedFeeCap = max(
+		addPercentage(originalFeeCap, cfg.BumpPercent()),
+		new(big.Int).Add(originalFeeCap, cfg.BumpMin()),
+	)
```
Is this different math than `bumpFeePriceByPercentage`?

ExtremelySunnyYK replied:
thanks for the note, we can refactor them into `bumpFeePriceByPercentage`

aalu1418 commented on common/fee/types/config.go:
```diff
@@ -0,0 +1,21 @@
+	FeeCapDefault() *big.Int
+	LimitMultiplier() float32
+	PriceDefault() *big.Int
+	TipCapDefault() *big.Int
```
For the dynamic static estimator - maybe you would have to use the generic fee calculation twice for the FeeCap and the TipCap?

not sure if this would actually work, I would need a closer look at the FeeCap and TipCap logic, but just tossing out an idea

aalu1418 commented on core/chains/evm/txmgr/broadcaster_test.go:
```diff
@@ -1486,7 +1487,7 @@ func TestEthBroadcaster_ProcessUnstartedEthTxs_Errors(t *testing.T) {
 		retryable, err := eb2.ProcessUnstartedTxs(testutils.Context(t), fromAddress)
 		require.Error(t, err)
-		require.Contains(t, err.Error(), "bumped gas price of 20 gwei is equal to original gas price of 20 gwei. ACTION REQUIRED: This is a configuration error, you must increase either EVM.GasEstimator.BumpPercent or EVM.GasEstimator.BumpMin")
+		require.Contains(t, err.Error(), "bumped fee price of 20000000000 fee unit is equal to original fee price of 20000000000 fee unit. ACTION REQUIRED: This is a configuration error, you must increase either FeeEstimator.BumpPercent or FeeEstimator.BumpMin")
```
is there a way to still return `wei` or `gwei` instead of `fee unit` in the errors? like a chain specific way for stringifying the value?

just thinking it's much more readable with proper units and can also inform the user which chain is being used

aalu1418 replied:
maybe something like passing a `FeeUnitToChainUnit` func for printing?

ExtremelySunnyYK replied:
good point, thanks Aaron for this. I think that having a chain specific way to stringify the values would be beneficial to the users. 

```
FeeUnitToChainUnit
		- Pass in chain type as a enum 
		- Based on the enum, it will update the units for printing based on big.Int
```

ExtremelySunnyYK commented on common/fee/models.go:
```diff
@@ -0,0 +1,183 @@
+}
+
+// CalculateFee computes the fee price and chain specific fee limit for a transaction.
+func CalculateFee(cfg feetypes.FixedPriceEstimatorConfig, bumpCfg feetypes.BumpConfig, feeLimit uint32, maxFeePrice *big.Int) (feePrice *big.Int, chainSpecificFeeLimit uint32, err error) {
```
I initially thought that the config would be chain agnostic. However looking back now i think that passing values around  would be much simpler and more composable so that the other chains would not have to implement its own interface. Thanks for the note!

ExtremelySunnyYK commented on common/fee/types/config.go:
```diff
@@ -0,0 +1,21 @@
+	PriceMax() *big.Int
+	BumpPercent() uint16
+	BumpMin() *big.Int
+	TipCapDefault() *big.Int
```
Thinking about it, i agree that tip cap is not the most chain agnostic concept. Based on @jmank88 's comment, changing config interfaces parameters out to values could help with this as well. 

I might be missing something here, but how does the evm specific calculator get passed into the generic static fee estimator? or would the legacy static fee estimator be calling the generic static fee estimator?

thanks for the digram btw!

ExtremelySunnyYK commented:
Sorry for the delay, made sure to double-check the logic to ensure the refactor logic was right.

Made significant changes to PR based on the comments:
   - General indentation/readability refactoring. thanks @jmank88 for the feedback
   - Removed Dynamic fee estimator.
   - Removed the use of config values as they aren't generalizable and not suitable for low-level math functions as Jordan pointed out.
   - Refactored names of functions to better reflect their purpose.
   - Added a basic `FeeUnitToChainUnit` function to convert fee values to native units.
   		- the list of chains will expand in the future.

jmank88 commented on common/fee/models.go:
```diff
@@ -0,0 +1,102 @@
+	feeLimit uint32,
+	maxFeePrice, defaultPrice, maxBumpPrice *big.Int,
+	bumpLimitMultiplier float32,
+) (feePrice *big.Int, chainSpecificFeeLimit uint32, err error) {
```
Unused error return arg?
```suggestion
) (feePrice *big.Int, chainSpecificFeeLimit uint32) {
```

ExtremelySunnyYK replied:
thanks for bringing this up, i think it will be better to return a nil instead in the caller function

jmank88 commented on core/chains/evm/gas/fixed_price_estimator.go:
```diff
@@ -89,6 +107,28 @@ func (f *fixedPriceEstimator) GetDynamicFee(_ context.Context, originalGasLimit
-func (f *fixedPriceEstimator) BumpDynamicFee(_ context.Context, originalFee DynamicFee, originalGasLimit uint32, maxGasPriceWei *assets.Wei, _ []EvmPriorAttempt) (bumped DynamicFee, chainSpecificGasLimit uint32, err error) {
-	return BumpDynamicFeeOnly(f.config, f.bhConfig.EIP1559FeeCapBufferBlocks(), f.lggr, f.config.TipCapDefault(), nil, originalFee, originalGasLimit, maxGasPriceWei)
+func (f *fixedPriceEstimator) BumpDynamicFee(
+	_ context.Context,
```
Why do we pass around unused context to these methods? Should we consider just dropping it?

aalu1418 replied:
i think this is conforming to the standard EVM estimator interface. other estimators like block history or L2 make calls to external services if i remember correctly

jmank88 commented on common/fee/utils.go:
```diff
@@ -1,23 +1,55 @@
 	return bigmath.Min(calculatedFeePrice, maxFeePrice), chainSpecificFeeLimit
 }
 
+func FeePriceLimit(userSpecifiedMax, maxFee *big.Int) *big.Int {
+	return bigmath.Min(userSpecifiedMax, maxFee)
```
If `FeePriceLimit` is just doing `min`, what value does the helper func add? Isn't that just more difficult to read?

ExtremelySunnyYK replied:
thought that it added a layer of abstraction, but i agree with you. It makes the code more readable. Thanks for bringing that up, not sure why we had that function initially but lets get rid of that function.

jmank88 commented on common/fee/utils.go:
```diff
@@ -1,23 +1,55 @@
+		return fmt.Sprintf("%s wei", fee.String())
+	case types.SOL:
+		return fmt.Sprintf("%s lamports", fee.String())
+	default:
+		return fee.String()
```
Do we normally tolerate unknown chain families? Or should we consider panicing here?

jmank88 commented on core/chains/evm/gas/fixed_price_estimator.go:
```diff
@@ -50,20 +56,32 @@ func (f *fixedPriceEstimator) Start(context.Context) error {
-	gasPrice = f.config.PriceDefault()
-	gasPrice, chainSpecificGasLimit = capGasPrice(gasPrice, maxGasPriceWei, f.config.PriceMax(), gasLimit, f.config.LimitMultiplier())
-	return
+func (f *fixedPriceEstimator) GetLegacyGas(_ context.Context, _ []byte, gasLimit uint32, maxGasPriceWei *assets.Wei, _ ...feetypes.Opt) (*assets.Wei, uint32, error) {
```
What is `_ []byte`? Are we stuck including that to satisfy an interface?

jmank88 commented on core/chains/evm/gas/fixed_price_estimator.go:
```diff
@@ -50,20 +56,32 @@ func (f *fixedPriceEstimator) Start(context.Context) error {
+	originalGasPrice *assets.Wei,
+	originalGasLimit uint32,
+	maxGasPriceWei *assets.Wei,
+	_ []EvmPriorAttempt,
```
Can we drop any of these unused args?
```suggestion
```

ExtremelySunnyYK replied:
I am looking into how we can drop this unused args. Right now they are used to satisfy interface, but not sure if we can drop them 🤔

aalu1418 commented on core/chains/evm/gas/fixed_price_estimator.go:
```diff
@@ -89,6 +107,28 @@ func (f *fixedPriceEstimator) GetDynamicFee(_ context.Context, originalGasLimit
+	)
 }
+
+func (f *fixedPriceEstimator) Name() string                                          { return f.lggr.Name() }
+func (f *fixedPriceEstimator) Ready() error                                          { return nil }
+func (f *fixedPriceEstimator) HealthReport() map[string]error                        { return map[string]error{} }
+func (f *fixedPriceEstimator) Close() error                                          { return nil }
+func (f *fixedPriceEstimator) OnNewLongestChain(_ context.Context, _ *evmtypes.Head) {}
```
nit: just wondering, why all the reorganization of the code?
it makes it seem like your PR is bigger than it actually is 🤷🏼

ExtremelySunnyYK replied:
I felt that the fee estimator was structured in a pretty confusing way. Felt that way when i was generalising the fee estimator initially, but now as well 🙃 

This style where we have the non essential methods at the bottom is used by the headtracker as well https://github.com/smartcontractkit/chainlink/blob/e42536d934e8c25a1b95d1baf2a2702cd7429595/core/chains/evm/headtracker/head_tracker.go#L50

ExtremelySunnyYK replied:
also expanded the other functions for readability due to long input parameters which can be confusing with named return values

ExtremelySunnyYK commented on common/fee/utils.go:
```diff
@@ -1,23 +1,55 @@
+}
+
+// FeeUnitToChainUnit returns the string representation of the given big.Int value in chain units.
+func FeeUnitToChainUnit(fee *big.Int, chain string) string {
```
This looks way cleaner, and i have not really seen this in `go`, thanks for suggesting this approach!

ExtremelySunnyYK commented on common/fee/models.go:
```diff
@@ -0,0 +1,102 @@
+	if err != nil {
+		return nil, 0, err
+	}
+	chainSpecificFeeLimit := ApplyMultiplier(originalFeeLimit, bumpLimitMultiplier)
```
that can also reduce function parameters as well, which is getting pretty long and can be hard to read.

jmank88 replied:
Yeah if some params were shared then it would make more sense to me. Maybe an earlier version just passed a config with all of these bundled? :shrug:

jmank88 commented on common/fee/models.go:
```diff
@@ -0,0 +1,100 @@
+	bumpPercent uint16,
+	toChainUnit func(*big.Int) string,
+) (*big.Int, error) {
+	feePrice, err := bumpFeePrice(lggr, currentFeePrice, originalFeePrice, maxFeePrice, maxBumpPrice, bumpMin, bumpPercent, toChainUnit)
+	if err != nil {
+		return nil, err
+	}
+	return feePrice, nil
```
Could return directly?
```suggestion
	return bumpFeePrice(lggr, currentFeePrice, originalFeePrice, maxFeePrice, maxBumpPrice, bumpMin, bumpPercent, toChainUnit)
```
Is there more room for simplification here? Do we actually need two different funcs?

jmank88 commented on common/fee/models.go:
```diff
@@ -0,0 +1,100 @@
+	return err != nil && (errors.Is(err, ErrBumpFeeExceedsLimit) || errors.Is(err, ErrBump) || errors.Is(err, ErrConnectivity))
+}
+
+// CalculateFee computes the fee price and chain specific fee limit for a transaction.
+func CalculateFee(
+	maxFeePrice, defaultPrice, maxBumpPrice *big.Int,
+) (feePrice *big.Int) {
+	feePrice = CapFeePrice(defaultPrice, maxFeePrice, maxBumpPrice)
+	return
```
Any more simplification available here?

ExtremelySunnyYK replied:
thanks for the catches, will look for other simplifications as well

aalu1418 commented on common/fee/utils.go:
```diff
@@ -8,16 +8,32 @@ import (
 }
+
+// Max returns the maximum of the two given big.Int values.
+func max(x, y *big.Int) *big.Int {
+	if x.Cmp(y) == 1 {
+		return x
+	}
+	return y
+}
```
and depending on that, we could potentially move the `addPercentage` into the big int utils or not

ExtremelySunnyYK replied:
i am thinking that if we were to move big.int into chainlink-relay, we could remove the `max` function. Also adding `addPercentage function. 

The reason why i copied over was mostly due to avoiding imports from core, and that seemed like a function that we didnt have to import a whole package over?

aalu1418 replied:
yeah that makes sense to me - eventually we'd have to untangle it. can you add a comment that this `Max` is a duplicate?

ExtremelySunnyYK replied:
sure! thank you for the note, forgot about leaving the comment.

aalu1418 commented on common/fee/utils.go:
```diff
@@ -3,21 +3,20 @@ package fee
 
-func GetMaxFeePrice(userSpecifiedMax, maxFee *big.Int) *big.Int {
-	return bigmath.Min(userSpecifiedMax, maxFee)
+func ApplyMultiplier(feeLimit uint32, multiplier float32) uint32 {
+	return uint32(decimal.NewFromBigInt(big.NewInt(0).SetUint64(uint64(feeLimit)), 0).Mul(decimal.NewFromFloat32(multiplier)).IntPart())
```
just wondering: would we ever be concerned this overflows?

my guess is unlikely since this is a multiplier on the individual fee units (gas, compute units) used but technically not impossible if a chain decides to have a huge block size

ExtremelySunnyYK replied:
i think thats a good point. 

If the product exceeds 4,294,967,295, then there would be an integer overflow when we cast it back to uint32. 

Should feeLimit be big.Int instead of uint32 then? I think the minimum would be to add checks and error handling if its more than that amount. 

Wondering if across chain, the limit would be more than `4,294,967,295` 🤔

aalu1418 reviewed (commented):
overall looks good - just curious about one thing

jmank88 commented on core/assets/wei.go:
```diff
@@ -76,6 +76,11 @@ func NewWeiI[T constraints.Signed](i T) *Wei {
 }
 
+// Returns the Wei value of the input big.Int.
+func NewWeiString(i *big.Int) string {
```
Funcs like this are typically called something like `FormatWei`

ExtremelySunnyYK replied:
thanks for the suggestion. Will update the name
